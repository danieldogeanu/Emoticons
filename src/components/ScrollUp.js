import React, { Component } from 'react';
import Icon from '../elements/Icon';
import '../styles/components/ScrollUp.scss';

class ScrollUp extends Component {
	constructor() {
		super();
		this.scrollUpBtn = React.createRef();
	}

	handleKeyUp = (e) => {
		if (e.keyCode === 84 && !this.SearchBarInputCL.contains('focused')) { // T
			this.scrollToTop();
		}
	}

	handleScroll = () => {
		let currentScroll = this.SimpleBarContentWrapper.scrollTop;
		let thisBtn = this.scrollUpBtn.current.classList;
		(currentScroll > 200) ? thisBtn.add('show') : thisBtn.remove('show');
	}

	scrollToTop = () => {
		let currentScroll = this.SimpleBarContentWrapper.scrollTop;
		if (currentScroll > 0) {
			window.requestAnimationFrame(this.scrollToTop);
			this.SimpleBarContentWrapper.scrollTo(0, currentScroll - (currentScroll/5));
		}
	}

	componentDidMount() {
		this.SimpleBarContentWrapper = document.querySelector('.simplebar-content-wrapper');
		this.SearchBarInputCL = document.querySelector('.SearchBar input[type="text"]').classList;
		this.SimpleBarContentWrapper.addEventListener('scroll', this.handleScroll);
		window.addEventListener('keyup', this.handleKeyUp);
	}

	componentWillUnmount() {
		this.SimpleBarContentWrapper.removeEventListener('scroll', this.handleScroll);
		window.removeEventListener('keyup', this.handleKeyUp);
	}

	render() {
		const scrollText = 'Scroll Back Up';
		return (
			<button className="ScrollUp"
				title={scrollText}
				onClick={this.scrollToTop}
				ref={this.scrollUpBtn}>
				<span className="show-for-screen-reader">{scrollText}</span>
				<Icon name="chevron-up" />
			</button>
		);
	}
}

export default ScrollUp;
