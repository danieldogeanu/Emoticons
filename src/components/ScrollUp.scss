@import '../styles/includes';

.ScrollUp {
	opacity: 0;
	display: block;
	position: fixed;
	right: rem(40);
	bottom: rem(40);
	width: rem(40);
	height: rem(40);
	border: rem(2) solid color-scheme(scrollup, normal);
	border-radius: rem(4);
	background-color: transparent;
	color: color-scheme(scrollup, normal);
	transition: border 0.3s ease, color 0.3s ease, opacity 0.5s ease;
	&:hover {
		border: rem(2) solid color-scheme(scrollup, hover);
		color: color-scheme(scrollup, hover);
	}
	&:focus, &:active { outline: none; }
	&.show { opacity: 1; }
}