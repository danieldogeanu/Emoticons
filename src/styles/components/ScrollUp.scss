@import '../includes';

.ScrollUp {
	opacity: 0;
	display: none;
	position: fixed;
	right: rem(40);
	bottom: rem(40);
	width: rem(40);
	height: rem(40);
	border: rem(2) solid color-scheme(scrollup, normal);
	border-radius: rem(4);
	background-color: transparent;
	color: color-scheme(scrollup, normal);
	transition: border 0.3s ease, color 0.3s ease, opacity 0.5s ease;
	@include breakpoint(min, rem(961)) { display: block; }
	&:hover {
		border: rem(2) solid color-scheme(scrollup, hover);
		color: color-scheme(scrollup, hover);
	}
	&:focus, &:active { outline: none; }
	&.show { opacity: 1; }
}

// TODO: Maybe include ScrollUp button on a black bar at the bottom of the screen on mobile along with ShortcutsButton.