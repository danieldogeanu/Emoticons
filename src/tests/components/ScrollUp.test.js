import React from 'react';
import {render} from '@testing-library/react';
import ScrollUp from '../../components/ScrollUp';

const scrollText = 'Scroll Back Up';
const compNames = {
	search: {
		bar: 'SearchBar',
		input: 'SearchInput',
	},
	screenReaderText: 'ScreenReaderText',
	scrollUp: 'ScrollUp',
};
const compClasses = {
	simplebar: 'simplebar-content-wrapper',
	list: 'List',
};

const SearchBar = () => (
	<form className={compNames.search.bar}
		data-testid={compNames.search.bar}>
		<input type="text" data-testid={compNames.search.input} />
	</form>
);

const ListContainer = () => (
	<div className={[
		compClasses.list,
		compClasses.simplebar,
	].join(' ')}></div>
);

const compShell = (component) => (
	<div data-testid="test-parent">
		<SearchBar />
		<ListContainer />
		{component}
	</div>
);

describe('ScrollUp Component', () => {

	it('renders scroll up button properly', () => {
		const {getByTitle, getByTestId, container} = render(compShell(<ScrollUp />));
		const renderedScrollBtn = getByTitle(scrollText);
		const renderedSRText = getByTestId(compNames.screenReaderText);
		const renderedIcon = container.querySelector('svg');

		expect(renderedScrollBtn).toBeInTheDocument();
		expect(renderedScrollBtn).toHaveClass(compNames.scrollUp);
		expect(renderedScrollBtn).toHaveAttribute('title', scrollText);
		expect(renderedScrollBtn).toContainElement(renderedSRText);
		expect(renderedScrollBtn).toContainElement(renderedIcon);
		expect(renderedScrollBtn).toMatchSnapshot();

		expect(renderedSRText).toHaveTextContent(scrollText);
		expect(renderedIcon).toHaveClass('icon', 'chevron-up');
	});

});
