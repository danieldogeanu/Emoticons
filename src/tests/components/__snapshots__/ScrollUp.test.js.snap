// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollUp Component renders scroll up button properly 1`] = `
<button
  class="ScrollUp"
  title="Scroll Back Up"
>
  <span
    class="show-for-screen-reader"
    data-testid="ScreenReaderText"
  >
    Scroll Back Up
  </span>
  <svg
    class="icon chevron-up"
  >
    <use
      xlink:href="#chevron-up"
    />
  </svg>
</button>
`;
